feat(site): add task notifications to user settings

ssncferreira opened the pull request:
## Description

Add UI support for the new `Task Events` notification templates (`Task Working` and `Task Idle`) added in backend PR https://github.com/coder/coder/pull/19965. The `/api/v2/notifications/templates/system` endpoint now returns these templates, therefore this PR keeps the UI in sync.

## Changes

* Added `Task Working` and `Task Idle` to notification template mocks so Storybook reflects backend data.
* Deployment Settings already lists all templates from the API, so no code change was needed (stories updated automatically).
* Updated the User Settings filter to include `Task Events` so they appear for end users.

<img width="2930" height="1666" alt="Screenshot 2025-09-29 at 17 06 22" src="https://github.com/user-attachments/assets/045d5dd7-13a4-4378-9bf3-9fd86a86e64b" />

Depends on: https://github.com/coder/coder/pull/19965

ssncferreira committed:
feat(site): add task notifications to user settings

buenos-nachos reviewed (commented):
I think all the changes look pretty straightforward and make sense

I'm going to approve in a minute – I just need to test something out with the story setup. I think we can sneak in some updates to all the `findByText` calls to make them less fragile and also test for accessibility

Parkreiner commented:
Okay, I did some digging, and I think the current testing setup is papering over some problems with the code

Namely, the change to the runtime code was in user settings version of `NotificationPage`, while the test update was in the deployment settings version of the `NotificationsPage`. I'm guessing that was an accident, but the test still succeeded, because we're only checking for text, not for whether the UI elements are valid

I think that realistically, we can keep the test updates for the deployment settings version. And then for the user settings version, we can get away with throwing this `play` function on the default story:
```tsx
play: async ({ canvasElement }) => {
	const canvas = within(canvasElement);

	// We only need to findBy for the first group, because all groups should
	// load in at the same time
	await canvas.findByRole("checkbox", { name: "Task Events" });
	void canvas.getByRole("checkbox", { name: "Template Events" });
	void canvas.getByRole("checkbox", { name: "User Events" });
	void canvas.getByRole("checkbox", { name: "Workspace Events" });
	void canvas.getByRole("checkbox", { name: "Custom Events" });
},
```

aslilac commented:
@Parkreiner why would you `void` the `get`s?

I'm not a fan of these kinds of patterns where you do something once and then do basically the same thing a different way many more times. it needs a comment to explain itself and doesn't provide any real benefit imo.

ssncferreira commented:
> Namely, the change to the runtime code was in user settings version of `NotificationPage`, while the test update was in the deployment settings version of the `NotificationsPage`. I'm guessing that was an accident, but the test still succeeded, because we're only checking for text, not for whether the UI elements are valid

@Parkreiner PR https://github.com/coder/coder/pull/19965 introduced a new `Task Events` notification with two templates, `Task Idle` and `Task Working`. As a result, the `/api/v2/notifications/templates/system` endpoint now returns these templates. To keep the UI consistent, I added the two templates to our mocks so Storybook stays in sync with the backend. The Deployment Settings view already renders all templates from the API, so no code change was required there and the stories updated automatically. For the User Settings view, we filter to settings a user can control, so I updated the filter to ensure the new `Task Events` appear as expected. I should have included this context in the PR description, sorry for the confusion.

ssncferreira committed:
chore: address comments

ssncferreira committed:
Merge remote-tracking branch 'origin/main' into ssncferreira/feat-site-task-notifications

ssncferreira commented:
> I think that realistically, we can keep the test updates for the deployment settings version. And then for the user settings version, we can get away with throwing this `play` function on the default story:
> 
> ```tsx
> play: async ({ canvasElement }) => {
> 	const canvas = within(canvasElement);
> 
> 	// We only need to findBy for the first group, because all groups should
> 	// load in at the same time
> 	await canvas.findByRole("checkbox", { name: "Task Events" });
> 	void canvas.getByRole("checkbox", { name: "Template Events" });
> 	void canvas.getByRole("checkbox", { name: "User Events" });
> 	void canvas.getByRole("checkbox", { name: "Workspace Events" });
> 	void canvas.getByRole("checkbox", { name: "Custom Events" });
> },
> ```
> 
> Trying to abstract that into a loop seemed like overkill

@Parkreiner addressed in https://github.com/coder/coder/pull/20006/commits/d1d600fda5d8c650225a8356895e25aa48c1236c
Added this test to the default story on the user’s notifications page. I used `findByRole` for all checkboxes because it makes it explicit that a checkbox exists for every event. Additionally, in Storybook’s UI it’s visible that all checks passed.
<img width="2992" height="3048" alt="Screenshot 2025-09-30 at 10 38 54" src="https://github.com/user-attachments/assets/b0a2c41d-104a-4b55-ac7f-7c5f7adacc65" />

ssncferreira commented:
@aslilac @Parkreiner thank you for the review and comments. I believe I’ve addressed all of them. Would it be okay if I merge the PR as-is? I just want to make sure it goes in before the release freeze 🙂 
If anything needs to change in the tests, would it be okay if I handle it in a follow-up PR?

buenos-nachos reviewed (commented):
Almost there. I think we accidentally added some flaky behavior for CI, but I just posted a suggestion for how to take care of it

ssncferreira committed:
fix: address comments

ssncferreira committed:
Merge remote-tracking branch 'origin/main' into ssncferreira/feat-site-task-notifications

ssncferreira commented on site/src/pages/UserSettingsPage/NotificationsPage/NotificationsPage.stories.tsx:
```diff
@@ -54,7 +54,20 @@ const meta = {
+		void canvas.findByRole("checkbox", { name: "Workspace Events" });
+
+		// Custom notification template
+		void canvas.findByRole("checkbox", { name: "Custom Events" });
```
Addressed in https://github.com/coder/coder/pull/20006/commits/fe720e39ec67f52497e43fe39f191391354d2d66
